frame/nomination-pools/src/lib.rs

```diff
@@ -582,25 +583,259 @@ pub struct PoolRoles<AccountId> {
+	pub max: Option<Perbill>,
+	/// Optional configuration around how often commission can be updated, and when the last
+	/// commission update took place.
+	pub throttle: Option<CommissionThrottle<T>>,
+}
+
+impl<T: Config> Commission<T> {
+	/// Get the current commission percentage of this pool.
+	///
+	/// Returns zero if commission has not yet been configured.
+	fn commission_or_zero(&self) -> Perbill {
+		self.current.as_ref().map(|(x, _)| *x).unwrap_or(Perbill::zero())
+	}
+
+	/// Returns true if a commission percentage updating to `to` would exhaust the throttle limit.
+	///
+	/// A commission update will be throttled (disallowed) if:
+	/// 1. not enough blocks have passed since the previous commission update took place, and
+	/// 2. the new commission is larger than the maximum allowed increase.
+	///
+	/// Throttling is not applied to commission updates if `current` is still `None`.
+	fn throttling(&self, to: &Perbill) -> bool {
+		if let Some(t) = self.throttle.as_ref() {
+			// factor previously updated block into whether user is throttled.
+			if t.previous_set_at.map_or(
+				// if no `previous_set_at` is set, throttled if the attempted increase
+				// in commission is greater than `max_increase`.
+				(*to).saturating_sub(self.commission_or_zero()) > t.change_rate.max_increase,
+				|p| {
+					// `min_delay` blocks must have been surpassed since last update.
+					if <frame_system::Pallet<T>>::block_number().saturating_sub(p) <
+						t.change_rate.min_delay
+					{
+						return true
+					}
+					// ensure the `max_increase` durations surpassed since the previous
+					// commission update allow the attempted commission increase.
+					//
+					// the attempted increase in commission relative to the current commission.
+					let attempted_increase = (*to).saturating_sub(self.commission_or_zero());
+
+					// the total durations passed since the last commission update.
+					let intervals_passed = self.intervals_since_block(&p, &t.change_rate.min_delay);
+
+					// the maximum allowed increase, where the current `max_increase`
+					// Perbill is converted into a u32 by multiplying itself with 100_u32
+					// and multiplied by durations passed, before being converted back
+					// into a Perbill.
+					let max_allowed_increase = Perbill::from_percent(
+						(t.change_rate.max_increase * 100_u32) * intervals_passed,
+					);
+
+					// throttled (true) if attempted increase is greater than the maximum
+					// allowed increase, as a percentage.
+					attempted_increase > max_allowed_increase
+				},
+			) {
+				return true
+			}
+		}
+		false
+	}
+
+	/// Set the pool's commission.
+	///
+	/// Update commission based on `commission` and `payee`. Do not allow a
+	/// commission above global maximum if set. If throttle is present, record the
+	///  current block as the previously updated commission. If the supplied
+	/// commission is zero, `None` will be inserted and `payee` will be ignored.
+	fn maybe_update_current(
```

Comment:
nitpick but the `maybe` prefix is usually used for a function that might set some value to `Some` or clear it back to `None` (generally, `maybe` is associated with `Option`) which is not the case here.